HPCC-19246 WsStore implementation

rpastrana opened the pull request:
- Provides espstore interface
- Provides dali based espstore implementation
- Provides ws_store esp service definition and implementation
- Includes ws_store instance in eclwatch
- Provides default Store for general eclwatch use
- Declares WsStoreAccess feature level flag and restricts use on all methods

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
@wangkx please review.

rpastrana commented on initfiles/componentfiles/configxml/@temp/esp_service_WsSMC.xsl:
```diff
@@ -125,6 +125,46 @@ This is required by its binding with ESP service '<xsl:value-of select="$espServ
+            <xsl:element name="Stores">
+                <xsl:element name="Store">
+                    <xsl:attribute name="description">ECL Watch User Preferences Store</xsl:attribute>
+                    <xsl:attribute name="name">ECLWatchPreferences</xsl:attribute>
```
@GordonSmith I'm adding wsstore to wssmc by default and creating a top level store (and setting it as the default store). Let me know if the name "ECLWatchPreferences" is compatible w/ your intent for this store

GordonSmith replied:
@rpastrana HPCCClientPreferences would work better?

rpastrana replied:
@GordonSmith sure...

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: eb8918e0a43f86c0b882e6de2b5875b13c5a0761
Number of Error(s): 0
CMake Error at cmake_modules/commonSetup.cmake:476 (add_library):
Number of Error(s): 0
CMakeFiles/CMakeOutput.log copied into ../CMakeOutput-2019-01-31_15-57-34.log
CMakeFiles/CMakeError.log copied into ../CMakeError-2019-01-31_15-57-34.log

[Build error:](https://gist.github.com/4df0e56c5741ae3722a4120aa619f7a0)
Build log [Build log](https://gist.github.com/HPCCSmoketest/4df0e56c5741ae3722a4120aa619f7a0#file-result-19-01-31-15-57-09-log)

wangkx reviewed (commented):
@rpastrana please see my comments.

rpastrana committed:
HPCC-19246 WsStore implementation

- Provides espstore interface
- Provides dali based espstore implementation
- Provides ws_store esp service definition and implementation
- Includes ws_store instance in eclwatch
- Provides default Store for general eclwatch use
- Introduces concept of Namespace ownership (global or user specific)
- Declares WsStoreAccess feature level flag and restricts use on all methods

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please review

HPCCSmoketest commented:
This PR (12127) is in the 1/2 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b94e92bb2aa4e26f07ebe93df34f789b3c1409d1
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |813 |813 |0 |
|test (thor) |739 |739 |0 |
|test (roxie) |887 |887 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  189 sec (00:03:09) |    0 sec (00:00:00) |    2 sec (00:00:02) |   17 sec (00:00:17) | 1435 sec (00:23:55) |   19 sec (00:00:19) | 1691 sec (00:28:11) |

wangkx reviewed (commented):
@rpastrana please see my comments.

rpastrana commented on esp/services/ws_store/ws_storeService.cpp:
```diff
@@ -0,0 +1,256 @@
+        Owned<IPropertyTreeIterator> iter = m_serviceConfig->getElements("Stores/Store");
+
+        StringBuffer owner;
+        owner.setf("%s/%s", _process, _service);
```
It might be odd to set the process/espinstance as the owner, but the store is created by the service. At this point there is no user involved.

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

rpastrana commented:
@wangkx please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 96ff252046246f06881525440ca5e74a2aef0617
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |744 |0 |
|test (roxie) |893 |893 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  189 sec (00:03:09) |    0 sec (00:00:00) |    2 sec (00:00:02) |   17 sec (00:00:17) | 1451 sec (00:24:11) |   20 sec (00:00:20) | 1705 sec (00:28:25) |

rpastrana commented on esp/services/ws_store/espstorelib/espstorebase.hpp:
```diff
@@ -0,0 +1,111 @@
+
+#define SDS_LOCK_TIMEOUT_KVSTORE (30*1000)
+
+interface IEspStore : extends IInterface
```
@RussWhitehead I'd like your input on the interface. thanks.

wangkx reviewed (commented):
The changes look good.

RussWhitehead commented on esp/services/ws_store/espstorelib/espstorebase.hpp:
```diff
@@ -0,0 +1,111 @@
+    virtual StringArray fetchKeySet(const char * storename, const char * ns, ISecUser * user, bool global) = 0;
+    virtual const char * fetch(const char * storename, const char * ns, const char * key, StringBuffer & value, ISecUser * username, bool global) = 0;
+    virtual IPropertyTree * fetchAll(const char * storename, const char * ns, ISecUser * user, bool global) = 0;
+    virtual StringArray fetchAllNamespaces(const char * storename, ISecUser * user, bool global) = 0;
```
As above, pass in ref to caller's StringArray

RussWhitehead commented on esp/services/ws_store/espstorelib/espstorebase.hpp:
```diff
@@ -0,0 +1,111 @@
+    virtual bool init(const char * name, const char * type, IPropertyTree * cfg) = 0;
+
+    virtual bool set(const char * storename, const char * thenamespace, const char * key, const char * value, ISecUser * owner, bool global) = 0;
+    virtual StringArray fetchKeySet(const char * storename, const char * ns, ISecUser * user, bool global) = 0;
```
Recommend having the user pass in a reference to a StringArray that it creates, instead of returning a reference to the service's StringArray

RussWhitehead commented on esp/services/ws_store/espstorelib/espstorebase.hpp:
```diff
@@ -0,0 +1,111 @@
+
+    virtual bool set(const char * storename, const char * thenamespace, const char * key, const char * value, ISecUser * owner, bool global) = 0;
+    virtual StringArray fetchKeySet(const char * storename, const char * ns, ISecUser * user, bool global) = 0;
+    virtual const char * fetch(const char * storename, const char * ns, const char * key, StringBuffer & value, ISecUser * username, bool global) = 0;
```
I think returning a pointer to something local to the called method is not safe, since another caller on another thread can change it, or worse yet, automatic data can go out of scope and be deleted

richardkchapman replied:
I think it actually returns a pointer to the supplied StringBuffer parameter (value). But our normal convention in such situations is to return StringBuffer &

rpastrana replied:
correct, it was returning pointer to the supplied StringBuf's buffer, removing return param, relying entirely on supplied StringBuff ref.

RussWhitehead reviewed (commented):
A few comments about returning pointers to locals

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9097f3fdb795d4007b8d1fcb8d6f2f7ada5399e0
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |744 |0 |
|test (roxie) |893 |893 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  190 sec (00:03:10) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1430 sec (00:23:50) |   19 sec (00:00:19) | 1683 sec (00:28:03) |

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

rpastrana commented:
@RussWhitehead please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5075b19650e8b42cf468aa01c21d73058d6afeac
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |744 |0 |
|test (roxie) |893 |893 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  195 sec (00:03:15) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1646 sec (00:27:26) |   19 sec (00:00:19) | 1898 sec (00:31:38) |

RussWhitehead commented on esp/services/ws_store/espstorelib/espstorebase.hpp:
```diff
@@ -39,17 +39,17 @@ interface IEspStore : extends IInterface
+    virtual IPropertyTree * getAllKeyProperties(const char * storename, const char * ns, const char * key, ISecUser * username, bool global) =0;
 };
 
 static const char* DALI_KVSTORE_PATH="/KVStore";
```
Consider removing implementation specific references (DALI) from the interface

RussWhitehead reviewed (commented):
Only reviewed the interface which looks good. One minor comment otherwise...

rpastrana commented:
@RussWhitehead please review.

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

RussWhitehead reviewed (approved):
@rpastrana Looks reasonable

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e192455ab369b6729ca103181f4c1b92725a643b
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |744 |0 |
|test (roxie) |893 |893 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1435 sec (00:23:55) |   19 sec (00:00:19) | 1689 sec (00:28:09) |

GordonSmith reviewed (changes requested):
One suggestion re "ECLWatchPreferences"

rpastrana commented:
@richardkchapman please merge

richardkchapman commented:
@rpastrana Please squash

HPCCSmoketest commented:
This PR (12127) is in the 6/12 position of the queue. Estimated start time is after ~3.0 hour(s)

HPCCSmoketest commented:
This PR (12127) is in the 5/11 position of the queue. Estimated start time is after ~2.5 hour(s)

HPCCSmoketest commented:
This PR (12127) is in the 4/10 position of the queue. Estimated start time is after ~2.0 hour(s)

HPCCSmoketest commented:
This PR (12127) is in the 3/9 position of the queue. Estimated start time is after ~1.5 hour(s)

rpastrana commented:
@richardkchapman squashed and ready to merge

HPCCSmoketest commented:
This PR (12127) is in the 2/8 position of the queue. Estimated start time is after ~1.0 hour(s)

HPCCSmoketest commented:
This PR (12127) is in the 1/7 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e192455ab369b6729ca103181f4c1b92725a643b
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |739 |5 |
|test (roxie) |893 |888 |5 |


[Errors:](https://gist.github.com/fe73a5b6b2595a5a2440d0d78ad3e8aa)
- test
  - thor
    - 597\. [spray_header_test (isTerminated=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_header_test-isterminated-false-w20190221-140648-3-log), 598\. [spray_header_test (isTerminated=true)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_header_test-isterminated-true-w20190221-140648-8-log), 605\. [spray_test_json (isSmallFile=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-false-w20190221-140650-6-log), 603\. [spray_test_json (isSmallFile=true, isUnBallanced=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-true,-isunballanced-false-w20190221-140650-9-log), 604\. [spray_test_json (isSmallFile=true, isUnBallanced=true)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-true,-isunballanced-true-w20190221-140650-1-log), 
  - roxie
    - 675\. [spray_header_test (isTerminated=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_header_test-isterminated-false-w20190221-141443-1-log), 676\. [spray_header_test (isTerminated=true)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_header_test-isterminated-true-w20190221-141443-5-log), 683\. [spray_test_json (isSmallFile=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-false-w20190221-141445-2-log), 681\. [spray_test_json (isSmallFile=true, isUnBallanced=false)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-true,-isunballanced-false-w20190221-141444-7-log), 682\. [spray_test_json (isSmallFile=true, isUnBallanced=true)](https://gist.github.com/HPCCSmoketest/fe73a5b6b2595a5a2440d0d78ad3e8aa#file-spray_test_json-issmallfile-true,-isunballanced-true-w20190221-141445-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  188 sec (00:03:08) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1422 sec (00:23:42) |   19 sec (00:00:19) | 1672 sec (00:27:52) |

AttilaVamos commented:
@AttilaVamos Those spray errors has been caused by my merged PR-https://github.com/hpcc-systems/HPCC-Platform/pull/12206 . I'm working on the fix. I am sorry about inconvenience caused.

HPCCSmoketest commented:
Process of PR-12127, label: HPCC-19246-wsstore starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a5a580c426bda5f264fc265ac399dd25e1b436f6
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |739 |5 |
|test (roxie) |893 |888 |5 |


[Errors:](https://gist.github.com/d248225df052a268269d69e815a6a95d)
- test
  - thor
    - 597\. [spray_header_test (isTerminated=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_header_test-isterminated-false-w20190221-163857-3-log), 598\. [spray_header_test (isTerminated=true)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_header_test-isterminated-true-w20190221-163857-5-log), 605\. [spray_test_json (isSmallFile=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-false-w20190221-163859-log), 603\. [spray_test_json (isSmallFile=true, isUnBallanced=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-true,-isunballanced-false-w20190221-163858-log), 604\. [spray_test_json (isSmallFile=true, isUnBallanced=true)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-true,-isunballanced-true-w20190221-163858-3-log), 
  - roxie
    - 675\. [spray_header_test (isTerminated=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_header_test-isterminated-false-w20190221-164641-4-log), 676\. [spray_header_test (isTerminated=true)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_header_test-isterminated-true-w20190221-164641-7-log), 683\. [spray_test_json (isSmallFile=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-false-w20190221-164643-1-log), 681\. [spray_test_json (isSmallFile=true, isUnBallanced=false)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-true,-isunballanced-false-w20190221-164643-log), 682\. [spray_test_json (isSmallFile=true, isUnBallanced=true)](https://gist.github.com/HPCCSmoketest/d248225df052a268269d69e815a6a95d#file-spray_test_json-issmallfile-true,-isunballanced-true-w20190221-164643-3-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  190 sec (00:03:10) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1459 sec (00:24:19) |   19 sec (00:00:19) | 1705 sec (00:28:25) |
